backend/src/handlers_prelude/github_handlers.rs

```diff
@@ -0,0 +1,236 @@
+use axum::{
+    extract::{State, Path},
+    http::StatusCode,
+    Json, Router,
+};
+use axum::routing::{get, post, put};
+use tracing::{error,info};
+use serde::{Serialize, Deserialize};
+use crate::gh::{get_all_branch_details, create_pull_request};
+use crate::handlers_prelude::eyre_to_axum_err;
+use crate::AppState;
+use color_eyre::Result;
+
+/// General API response structure
+#[derive(Serialize, Debug)]
+pub struct ApiResponse<T> {
+    pub status: String,
+    pub message: String,
+    pub data: Option<T>,
+}
+
+/// Error response structure
+#[derive(Serialize, Debug)]
+pub struct ApiErrorResponse {
+    pub error: String,
+}
+
+/// Represents the structure for a pull request creation response
+#[derive(Serialize, Debug)]
+pub struct CreatePRData {
+    pub pull_request_url: String,
+}
+
+/// Represents the structure for a branch listing response
+#[derive(Serialize, Debug)]
+pub struct BranchesData {
+    pub branches: Vec<String>,
+}
+
+/// Request structure for creating a pull request
+#[derive(Deserialize, Debug)]
+pub struct CreatePRRequest {
+    pub head_branch: String,
+    pub base_branch: String,
+    pub title: String,
+    pub description: String,
+}
+
+/// Retrieves the GitHub access token from the application state.
+///
+/// This asynchronous function accesses the application state to fetch
+/// the GitHub access token using the provided HTTP client. If the token
+/// retrieval fails, it returns an error with a corresponding HTTP status code
+/// and a descriptive error message.
+///
+/// # Parameters
+/// - `state`: A reference to the `AppState`, which contains the necessary
+///   credentials and the HTTP client for making the request.
+///
+/// # Returns
+/// - On success, returns a `Result` containing the GitHub access token as a `String`.
+/// - On failure, returns a tuple containing the appropriate `StatusCode` and an error message.
+///
+/// # Errors
+/// This function can fail if there are issues with the credentials store
+/// or if the token cannot be retrieved for any reason.
+async fn get_github_token(state: &AppState) -> Result<String, (StatusCode, String)> {
+    state.gh_credentials.get(&state.reqwest_client, &state.config.oauth.github.client_id).await.map_err(|err| {
+        eyre_to_axum_err(err)
+    })
+}
+
+/// Fetches the list of branches from a GitHub repository.
+///
+/// This function interacts with the GitHub API to retrieve all branches
+/// for a specific repository. It requires an access token for authentication.
+///
+/// # Parameters
+/// - `reqwest_client`: An instance of `reqwest::Client` used to make HTTP requests.
+/// - `token`: A string slice representing the GitHub access token used for authentication.
+///
+/// # Returns
+/// - On success, returns a `Result` containing a vector of branch names as strings.
+/// - On failure, returns an error that implements the `std::error::Error` trait, indicating
+///   what went wrong during the API request.
```

Comment:
I don't believe `(StatusCode, String)` implements `std::errror::Error`, that's not necessarily a guarantee with result, you can have a result of any two types.